src/main/java/services/SubmitService.java

```diff
@@ -4,23 +4,37 @@
 import org.springframework.http.HttpEntity;
 import org.springframework.http.HttpMethod;
 import org.springframework.http.ResponseEntity;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.web.client.DefaultResponseErrorHandler;
+import org.springframework.web.client.ResponseErrorHandler;
 import org.springframework.web.client.RestTemplate;
 
+import java.io.IOException;
+
 /**
  * Service that handles form submission to CommCareHQ
  */
-public class SubmitService {
+public class SubmitService extends DefaultResponseErrorHandler {
 
     @Autowired
     RestoreFactory restoreFactory;
 
     public ResponseEntity<String> submitForm(String formXml, String submitUrl) {
+        restoreFactory.setAutoCommit(false);
```

Comment:
I think these always need to be wrapped in try{} finally{} blocks

Reply:
Er, not the setAutoCommit(False) part, the part after (including the potential rollback). The general pattern of

```
setNotAutoCommit
try{
   dostuff
} fail {
rollback
} finally {
  autocommitbackon
}
```